File: code/backend/python/app/core/services/story/story_query_service.py
from typing import List, Optional, Dict, Any
from app.core.models.story.story_summary import StorySummary
from app.core.models.story.story_detail import StoryDetail
from app.core.models.story.difficulty_info import DifficultyInfo
from app.core.models.story.character_template_info import CharacterTemplateInfo
from app.core.models.story.class_info import ClassInfo
from app.core.models.story.class_bonus_info import ClassBonusInfo
from app.core.models.story.trait_info import TraitInfo
from app.core.models.story.card_info import CardInfo
from app.core.ports.story.story_query_port import StoryQueryPort
from app.core.ports.story.story_read_port import StoryReadPort

class StoryQueryService(StoryQueryPort):
    def __init__(self, read_port: StoryReadPort):
        self.read_port = read_port

    def list_public_stories(self, lang: str = "en") -> List[StorySummary]:
        lang = lang or "en"
        raw_stories = self.read_port.find_public_stories()
        return [self._map_to_summary(r, lang) for r in raw_stories]

    def list_all_stories(self, lang: str = "en") -> List[StorySummary]:
        lang = lang or "en"
        raw_stories = self.read_port.find_all_stories()
        return [self._map_to_summary(r, lang) for r in raw_stories]

    def list_categories(self) -> List[str]:
        return self.read_port.find_unique_categories()

    def list_groups(self) -> List[str]:
        return self.read_port.find_unique_groups()

    def list_stories_by_category(self, category: str, lang: str = "en") -> List[StorySummary]:
        lang = lang or "en"
        raw_stories = self.read_port.find_stories_by_category(category)
        return [self._map_to_summary(r, lang) for r in raw_stories]

    def list_stories_by_group(self, group: str, lang: str = "en") -> List[StorySummary]:
        lang = lang or "en"
        raw_stories = self.read_port.find_stories_by_group(group)
        return [self._map_to_summary(r, lang) for r in raw_stories]

    def get_story_detail(self, uuid: str, lang: str = "en") -> Optional[StoryDetail]:
        lang = lang or "en"
        raw_story = self.read_port.find_story_by_uuid(uuid)
        if not raw_story:
            return None

        story_id = raw_story.get("id")
        texts = self.read_port.find_texts_for_story(story_id)
        
        # Difficulty
        raw_diffs = self.read_port.find_difficulties_for_story(story_id)
        difficulties = []
        for rd in raw_diffs:
            desc = self._resolve_text(texts, rd.get("id_text_description"), lang)
            difficulties.append(DifficultyInfo(
                uuid=rd.get("uuid", ""),
                description=desc,
                expCost=rd.get("exp_cost", 5),
                maxWeight=rd.get("max_weight", 10),
                minCharacter=rd.get("min_character", 1),
                maxCharacter=rd.get("max_character", 4),
                costHelpComa=rd.get("cost_help_coma", 3),
                costMaxCharacteristics=rd.get("cost_max_characteristics", 3),
                numberMaxFreeAction=rd.get("number_max_free_action", 1),
                traitCostPositiveBudget=rd.get("trait_cost_positive_budget"),
                traitCostNegativeBudget=rd.get("trait_cost_negative_budget"),
                idCard=rd.get("id_card"),
                card=self._resolve_card(story_id, texts, rd.get("id_card"), lang),
                life=rd.get("life", 100),
                energy=rd.get("energy", 100),
                sad=rd.get("sad", 0),
                dexterity=rd.get("dexterity", 10),
                intelligence=rd.get("intelligence", 10),
                constitution=rd.get("constitution", 10),
                weight=rd.get("weight", 10)
            ))

        # Counts
        loc_count = self.read_port.count_locations_for_story(story_id)
        event_count = self.read_port.count_events_for_story(story_id)
        item_count = self.read_port.count_items_for_story(story_id)

        title = self._resolve_text(texts, raw_story.get("id_text_title"), lang)
        desc = self._resolve_text(texts, raw_story.get("id_text_description"), lang)
        copyright_txt = self._resolve_text(texts, raw_story.get("id_text_copyright"), lang)

        # Step 15: Classes, Templates, Traits
        raw_classes = self.read_port.find_classes_for_story(story_id)
        raw_class_bonuses = self.read_port.find_class_bonuses_for_story(story_id)
        raw_templates = self.read_port.find_character_templates_for_story(story_id)
        raw_traits = self.read_port.find_traits_for_story(story_id)

        classes = []
        for c in raw_classes:
            cls_name = self._resolve_text(texts, c.get("id_text_name"), lang)
            cls_desc = self._resolve_text(texts, c.get("id_text_description"), lang)
            cls_id = c.get("id")
            cls_bonuses = []
            for b in raw_class_bonuses:
                if b.get("id_class") == cls_id:
                    cls_bonuses.append(ClassBonusInfo(
                        uuid=b.get("uuid") or str(b.get("id", "")),
                        statistic=b.get("statistic") or b.get("bonus_type"),
                        value=b.get("value") if b.get("value") is not None else (b.get("bonus_value") or 0),
                    ))
            classes.append(ClassInfo(
                uuid=c.get("uuid") or str(c.get("id", "")),
                id=c.get("id"),
                name=cls_name,
                description=cls_desc,
                weightMax=c.get("weight_max", 0) or 0,
                dexterityBase=c.get("dexterity_base", 0) or 0,
                intelligenceBase=c.get("intelligence_base", 0) or 0,
                constitutionBase=c.get("constitution_base", 0) or 0,
                idCard=c.get("id_card"),
                card=self._resolve_card(story_id, texts, c.get("id_card"), lang),
                bonuses=cls_bonuses,
            ))

        templates = []
        for t in raw_templates:
            tpl_name = self._resolve_text(texts, t.get("id_text_name"), lang)
            tpl_desc = self._resolve_text(texts, t.get("id_text_description"), lang)
            templates.append(CharacterTemplateInfo(
                uuid=t.get("uuid") or str(t.get("id", "")),
                name=tpl_name,
                description=tpl_desc,
                lifeMax=t.get("life_max", 0) or 0,
                energyMax=t.get("energy_max", 0) or 0,
                sadMax=t.get("sad_max", 0) or 0,
                dexterityStart=t.get("dexterity_start", 0) or 0,
                intelligenceStart=t.get("intelligence_start", 0) or 0,
                constitutionStart=t.get("constitution_start", 0) or 0,
                idCard=t.get("id_card"),
                card=self._resolve_card(story_id, texts, t.get("id_card"), lang),
                idClassPermitted=t.get("id_class_permitted"),
                idClassProhibited=t.get("id_class_prohibited"),
            ))

        traits = [self._to_trait_info(story_id, texts, tr, lang) for tr in raw_traits]

        # Step 15: Card
        card = None
        story_id_card = raw_story.get("id_card")
        if story_id_card is not None:
            raw_card = self.read_port.find_card_for_story(story_id, story_id_card)
            if raw_card:
                card_title = self._resolve_text(texts, raw_card.get("id_text_title") or raw_card.get("id_text_name"), lang)
                card_description = self._resolve_text(texts, raw_card.get("id_text_description"), lang)
                card_copyright_text = self._resolve_text(texts, raw_card.get("id_text_copyright"), lang)
                card = CardInfo(
                    uuid=raw_card.get("uuid") or str(raw_card.get("id", "")),
                    cardType=raw_card.get("card_type"),
                    urlImage=raw_card.get("url_image"),
                    alternativeImage=raw_card.get("alternative_image"),
                    awesomeIcon=raw_card.get("awesome_icon"),
                    styleMain=raw_card.get("style_main"),
                    styleDetail=raw_card.get("style_detail"),
                    styleImageLittle=raw_card.get("style_image_little"),
                    styleImageMedium=raw_card.get("style_image_medium"),
                    styleImageLarge=raw_card.get("style_image_large"),
                    title=card_title,
                    description=card_description,
                    copyrightText=card_copyright_text,
                    linkCopyright=raw_card.get("link_copyright")
                )

        return StoryDetail(
            uuid=raw_story.get("uuid"),
            title=title,
            description=desc,
            author=raw_story.get("author"),
            category=raw_story.get("category"),
            group=raw_story.get("group_name"),
            visibility=raw_story.get("visibility"),
            priority=raw_story.get("priority", 0),
            peghi=raw_story.get("peghi", 0),
            versionMin=raw_story.get("version_min"),
            versionMax=raw_story.get("version_max"),
            clockSingularDescription=self._resolve_text(texts, raw_story.get("id_text_clock_singular"), lang),
            clockPluralDescription=self._resolve_text(texts, raw_story.get("id_text_clock_plural"), lang),
            idTextClockSingular=raw_story.get("id_text_clock_singular"),
            idTextClockPlural=raw_story.get("id_text_clock_plural"),
            copyrightText=copyright_txt,
            linkCopyright=raw_story.get("link_copyright"),
            locationCount=loc_count,
            eventCount=event_count,
            itemCount=item_count,
            classCount=len(raw_classes),
            characterTemplateCount=len(raw_templates),
            traitCount=len(raw_traits),
            difficulties=difficulties,
            classes=classes,
            characterTemplates=templates,
            traits=traits,
            card=card
        )

    def _map_to_summary(self, raw_story: Dict[str, Any], lang: str) -> StorySummary:
        story_id = raw_story.get("id")
        texts = self.read_port.find_texts_for_story(story_id)
        
        title = self._resolve_text(texts, raw_story.get("id_text_title"), lang)
        desc = self._resolve_text(texts, raw_story.get("id_text_description"), lang)
        
        diff_count = len(self.read_port.find_difficulties_for_story(story_id))

        # Card resolution
        card = None
        story_id_card = raw_story.get("id_card")
        if story_id_card is not None:
            raw_card = self.read_port.find_card_for_story(story_id, story_id_card)
            if raw_card:
                card_title = self._resolve_text(texts, raw_card.get("id_text_title") or raw_card.get("id_text_name"), lang)
                card_description = self._resolve_text(texts, raw_card.get("id_text_description"), lang)
                card_copyright_text = self._resolve_text(texts, raw_card.get("id_text_copyright"), lang)
                card = CardInfo(
                    uuid=raw_card.get("uuid") or str(raw_card.get("id", "")),
                    cardType=raw_card.get("card_type"),
                    urlImage=raw_card.get("url_image"),
                    alternativeImage=raw_card.get("alternative_image"),
                    awesomeIcon=raw_card.get("awesome_icon"),
                    styleMain=raw_card.get("style_main"),
                    styleDetail=raw_card.get("style_detail"),
                    styleImageLittle=raw_card.get("style_image_little"),
                    styleImageMedium=raw_card.get("style_image_medium"),
                    styleImageLarge=raw_card.get("style_image_large"),
                    title=card_title,
                    description=card_description,
                    copyrightText=card_copyright_text,
                    linkCopyright=raw_card.get("link_copyright")
                )

        return StorySummary(
            uuid=raw_story.get("uuid"),
            title=title,
            description=desc,
            author=raw_story.get("author"),
            category=raw_story.get("category"),
            group=raw_story.get("group_name"),
            visibility=raw_story.get("visibility"),
            priority=raw_story.get("priority", 0),
            peghi=raw_story.get("peghi", 0),
            difficultyCount=diff_count,
            card=card,
            idTextClockSingular=raw_story.get("id_text_clock_singular"),
            idTextClockPlural=raw_story.get("id_text_clock_plural")
        )

    # === Step 23: Trait listing filtered by class ===

    def list_traits_for_class(self, story_uuid: str, class_uuid: str, lang: str = "en"):
        """Lists the story traits selectable with the given class.

        Returns a ``(status, traits)`` tuple where status is ``"OK"``,
        ``"STORY_NOT_FOUND"`` or ``"CLASS_NOT_FOUND"``.
        """
        if not story_uuid or not story_uuid.strip():
            return "STORY_NOT_FOUND", []
        raw_story = self.read_port.find_story_by_uuid(story_uuid)
        if not raw_story:
            return "STORY_NOT_FOUND", []
        story_id = raw_story["id"]

        clazz = None
        if class_uuid and class_uuid.strip():
            for c in self.read_port.find_classes_for_story(story_id):
                if c.get("uuid") == class_uuid:
                    clazz = c
                    break
        if clazz is None:
            return "CLASS_NOT_FOUND", []
        class_id = clazz.get("id")

        texts = self.read_port.find_texts_for_story(story_id)
        traits = [
            self._to_trait_info(story_id, texts, tr, lang)
            for tr in self.read_port.find_traits_for_story(story_id)
            if self._is_trait_selectable(tr, class_id)
        ]
        return "OK", traits

    @staticmethod
    def _is_trait_selectable(tr: Dict[str, Any], class_id) -> bool:
        permitted = tr.get("id_class_permitted")
        prohibited = tr.get("id_class_prohibited")
        permitted_ok = permitted is None or (class_id is not None and permitted == class_id)
        prohibited_ok = prohibited is None or class_id is None or prohibited != class_id
        return permitted_ok and prohibited_ok

    def _to_trait_info(self, story_id: int, texts: List[Dict[str, Any]],
                       tr: Dict[str, Any], lang: str) -> TraitInfo:
        tr_name = self._resolve_text(texts, tr.get("id_text_name"), lang)
        tr_desc = self._resolve_text(texts, tr.get("id_text_description"), lang)
        return TraitInfo(
            uuid=tr.get("uuid") or str(tr.get("id", "")),
            name=tr_name,
            description=tr_desc,
            costPositive=tr.get("cost_positive", 0) or 0,
            costNegative=tr.get("cost_negative", 0) or 0,
            idClassPermitted=tr.get("id_class_permitted"),
            idClassProhibited=tr.get("id_class_prohibited"),
            idCard=tr.get("id_card"),
            card=self._resolve_card(story_id, texts, tr.get("id_card"), lang),
            # v0.35.2 — reported on BOTH projections; nothing is filtered out here.
            hideOnStartMatch=tr.get("hide_on_start_match") == 1,
            life=tr.get("life", 0) or 0,
            energy=tr.get("energy", 0) or 0,
            sad=tr.get("sad", 0) or 0,
            dexterity=tr.get("dexterity", 0) or 0,
            intelligence=tr.get("intelligence", 0) or 0,
            constitution=tr.get("constitution", 0) or 0,
            weight=tr.get("weight", 0) or 0,
        )

    def _resolve_card(self, story_id: int, texts: List[Dict[str, Any]], id_card: Optional[int], lang: str) -> Optional[CardInfo]:
        if id_card is None:
            return None
        raw_card = self.read_port.find_card_for_story(story_id, id_card)
        if not raw_card:
            return None
        card_title = self._resolve_text(texts, raw_card.get("id_text_title") or raw_card.get("id_text_name"), lang)
        card_description = self._resolve_text(texts, raw_card.get("id_text_description"), lang)
        card_copyright_text = self._resolve_text(texts, raw_card.get("id_text_copyright"), lang)
        return CardInfo(
            uuid=raw_card.get("uuid") or str(raw_card.get("id", "")),
            cardType=raw_card.get("card_type"),
            urlImage=raw_card.get("url_image"),
            alternativeImage=raw_card.get("alternative_image"),
            awesomeIcon=raw_card.get("awesome_icon"),
            styleMain=raw_card.get("style_main"),
            styleDetail=raw_card.get("style_detail"),
            styleImageLittle=raw_card.get("style_image_little"),
            styleImageMedium=raw_card.get("style_image_medium"),
            styleImageLarge=raw_card.get("style_image_large"),
            title=card_title,
            description=card_description,
            copyrightText=card_copyright_text,
            linkCopyright=raw_card.get("link_copyright")
        )

    def _resolve_text(self, texts: List[Dict[str, Any]], txt_id: Optional[int], target_lang: str) -> Optional[str]:
        if txt_id is None:
            return None
            
        candidates = [t for t in texts if t.get("id_text") == txt_id]
        if not candidates:
            return None

        # 1. Exact language match
        for t in candidates:
            if t.get("lang") == target_lang:
                return t.get("short_text") or t.get("long_text")

        # 2. English fallback
        for t in candidates:
            if t.get("lang") == "en":
                return t.get("short_text") or t.get("long_text")

        # 3. Any available
        t = candidates[0]
        return t.get("short_text") or t.get("long_text")

File: code/backend/php/src/Core/Service/Story/StoryQueryService.php
<?php

declare(strict_types=1);

namespace Games\Paths\Core\Service\Story;

use Games\Paths\Core\Domain\Story\CardInfo;
use Games\Paths\Core\Domain\Story\CharacterTemplateInfo;
use Games\Paths\Core\Domain\Story\ClassInfo;
use Games\Paths\Core\Domain\Story\DifficultyInfo;
use Games\Paths\Core\Domain\Story\StoryDetail;
use Games\Paths\Core\Domain\Story\StorySummary;
use Games\Paths\Core\Domain\Story\TraitInfo;
use Games\Paths\Core\Port\Story\StoryQueryPort;
use Games\Paths\Core\Port\Story\StoryReadPort;

class StoryQueryService implements StoryQueryPort
{
    private StoryReadPort $readPort;

    public function __construct(StoryReadPort $readPort)
    {
        $this->readPort = $readPort;
    }

    public function listPublicStories(string $lang = 'en'): array
    {
        $rawStories = $this->readPort->findPublicStories();
        return array_map(fn($r) => $this->mapToSummary($r, $lang), $rawStories);
    }

    public function listAllStories(string $lang = 'en'): array
    {
        $rawStories = $this->readPort->findAllStories();
        return array_map(fn($r) => $this->mapToSummary($r, $lang), $rawStories);
    }

    public function listCategories(): array
    {
        return $this->readPort->findUniqueCategories();
    }

    public function listGroups(): array
    {
        return $this->readPort->findUniqueGroups();
    }

    public function listStoriesByCategory(string $category, string $lang = 'en'): array
    {
        $rawStories = $this->readPort->findStoriesByCategory($category);
        return array_map(fn($r) => $this->mapToSummary($r, $lang), $rawStories);
    }

    public function listStoriesByGroup(string $group, string $lang = 'en'): array
    {
        $rawStories = $this->readPort->findStoriesByGroup($group);
        return array_map(fn($r) => $this->mapToSummary($r, $lang), $rawStories);
    }

    public function getStoryDetail(string $uuid, string $lang = 'en'): ?StoryDetail
    {
        $rawStory = $this->readPort->findStoryByUuid($uuid);
        if (!$rawStory) {
            return null;
        }

        $storyId = (int) $rawStory['id'];
        $texts = $this->readPort->findTextsForStory($storyId);

        $rawDiffs = $this->readPort->findDifficultiesForStory($storyId);
        $difficulties = [];
        foreach ($rawDiffs as $rd) {
            $desc = $this->resolveText($texts, isset($rd['id_text_description']) ? (int)$rd['id_text_description'] : null, $lang);
            $difficulties[] = new DifficultyInfo(
                $rd['uuid'] ?? '',
                $desc,
                isset($rd['exp_cost']) ? (int)$rd['exp_cost'] : 5,
                isset($rd['max_weight']) ? (int)$rd['max_weight'] : 10,
                isset($rd['min_character']) ? (int)$rd['min_character'] : 1,
                isset($rd['max_character']) ? (int)$rd['max_character'] : 4,
                isset($rd['cost_help_coma']) ? (int)$rd['cost_help_coma'] : 3,
                isset($rd['cost_max_characteristics']) ? (int)$rd['cost_max_characteristics'] : 3,
                isset($rd['number_max_free_action']) ? (int)$rd['number_max_free_action'] : 1
            );
        }

        $locCount = $this->readPort->countLocationsForStory($storyId);
        $eventCount = $this->readPort->countEventsForStory($storyId);
        $itemCount = $this->readPort->countItemsForStory($storyId);

        $title = $this->resolveText($texts, isset($rawStory['id_text_title']) ? (int)$rawStory['id_text_title'] : null, $lang);
        $desc = $this->resolveText($texts, isset($rawStory['id_text_description']) ? (int)$rawStory['id_text_description'] : null, $lang);
        $copyrightTxt = $this->resolveText($texts, isset($rawStory['id_text_copyright']) ? (int)$rawStory['id_text_copyright'] : null, $lang);

        // Step 15: Classes, Templates, Traits
        $rawClasses = $this->readPort->findClassesForStory($storyId);
        $rawTemplates = $this->readPort->findCharacterTemplatesForStory($storyId);
        $rawTraits = $this->readPort->findTraitsForStory($storyId);

        $classes = [];
        foreach ($rawClasses as $c) {
            $clsName = $this->resolveText($texts, isset($c['id_text_name']) ? (int)$c['id_text_name'] : null, $lang);
            $clsDesc = $this->resolveText($texts, isset($c['id_text_description']) ? (int)$c['id_text_description'] : null, $lang);
            $classes[] = new ClassInfo(
                $c['uuid'] ?? (string)($c['id'] ?? ''),
                $clsName,
                $clsDesc,
                isset($c['weight_max']) ? (int)$c['weight_max'] : 0,
                isset($c['dexterity_base']) ? (int)$c['dexterity_base'] : 0,
                isset($c['intelligence_base']) ? (int)$c['intelligence_base'] : 0,
                isset($c['constitution_base']) ? (int)$c['constitution_base'] : 0
            );
        }

        $characterTemplates = [];
        foreach ($rawTemplates as $t) {
            $tplName = $this->resolveText($texts, isset($t['id_text_name']) ? (int)$t['id_text_name'] : null, $lang);
            $tplDesc = $this->resolveText($texts, isset($t['id_text_description']) ? (int)$t['id_text_description'] : null, $lang);
            $characterTemplates[] = new CharacterTemplateInfo(
                $t['uuid'] ?? (string)($t['id'] ?? ''),
                $tplName,
                $tplDesc,
                isset($t['life_max']) ? (int)$t['life_max'] : 0,
                isset($t['energy_max']) ? (int)$t['energy_max'] : 0,
                isset($t['sad_max']) ? (int)$t['sad_max'] : 0,
                isset($t['dexterity_start']) ? (int)$t['dexterity_start'] : 0,
                isset($t['intelligence_start']) ? (int)$t['intelligence_start'] : 0,
                isset($t['constitution_start']) ? (int)$t['constitution_start'] : 0
            );
        }

        $traits = [];
        foreach ($rawTraits as $tr) {
            $trName = $this->resolveText($texts, isset($tr['id_text_name']) ? (int)$tr['id_text_name'] : null, $lang);
            $trDesc = $this->resolveText($texts, isset($tr['id_text_description']) ? (int)$tr['id_text_description'] : null, $lang);
            $traits[] = new TraitInfo(
                $tr['uuid'] ?? (string)($tr['id'] ?? ''),
                $trName,
                $trDesc,
                isset($tr['cost_positive']) ? (int)$tr['cost_positive'] : 0,
                isset($tr['cost_negative']) ? (int)$tr['cost_negative'] : 0,
                isset($tr['id_class_permitted']) ? (int)$tr['id_class_permitted'] : null,
                isset($tr['id_class_prohibited']) ? (int)$tr['id_class_prohibited'] : null
            );
        }

        // Step 15: Card
        $card = null;
        $storyIdCard = $rawStory['id_card'] ?? null;
        if ($storyIdCard !== null) {
            $rawCard = $this->readPort->findCardForStory($storyId, (int)$storyIdCard);
            if ($rawCard) {
                $cardTitleTextId = isset($rawCard['id_text_title']) ? (int)$rawCard['id_text_title']
                    : (isset($rawCard['id_text_name']) ? (int)$rawCard['id_text_name'] : null);
                $cardTitle = $this->resolveText($texts, $cardTitleTextId, $lang);
                $cardDescTextId = isset($rawCard['id_text_description']) ? (int)$rawCard['id_text_description'] : null;
                $cardDescription = $this->resolveText($texts, $cardDescTextId, $lang);
                $cardCopyrightTextId = isset($rawCard['id_text_copyright']) ? (int)$rawCard['id_text_copyright'] : null;
                $cardCopyrightText = $this->resolveText($texts, $cardCopyrightTextId, $lang);
                $card = new CardInfo(
                    $rawCard['uuid'] ?? (string)($rawCard['id'] ?? ''),
                    $rawCard['image_url'] ?? null,
                    $rawCard['alternative_image'] ?? null,
                    $rawCard['awesome_icon'] ?? null,
                    $rawCard['style_main'] ?? null,
                    $rawCard['style_detail'] ?? null,
                    $cardTitle,
                    $cardDescription,
                    $cardCopyrightText,
                    $rawCard['link_copyright'] ?? null
                );
            }
        }

        return new StoryDetail(
            $rawStory['uuid'],
            $title,
            $desc,
            $rawStory['author'] ?? null,
            $rawStory['category'] ?? null,
            $rawStory['group_name'] ?? null,
            $rawStory['visibility'] ?? null,
            isset($rawStory['priority']) ? (int)$rawStory['priority'] : 0,
            isset($rawStory['peghi']) ? (int)$rawStory['peghi'] : 0,
            $rawStory['version_min'] ?? null,
            $rawStory['version_max'] ?? null,
            $rawStory['clock_singular'] ?? null,
            $rawStory['clock_plural'] ?? null,
            $copyrightTxt,
            $rawStory['link_copyright'] ?? null,
            $locCount,
            $eventCount,
            $itemCount,
            count($rawClasses),
            count($rawTemplates),
            count($rawTraits),
            $difficulties,
            $characterTemplates,
            $classes,
            $traits,
            $card
        );
    }

    private function mapToSummary(array $rawStory, string $lang): StorySummary
    {
        $storyId = (int) $rawStory['id'];
        $texts = $this->readPort->findTextsForStory($storyId);

        $title = $this->resolveText($texts, isset($rawStory['id_text_title']) ? (int)$rawStory['id_text_title'] : null, $lang);
        $desc = $this->resolveText($texts, isset($rawStory['id_text_description']) ? (int)$rawStory['id_text_description'] : null, $lang);
        
        $diffCount = count($this->readPort->findDifficultiesForStory($storyId));

        return new StorySummary(
            $rawStory['uuid'],
            $title,
            $desc,
            $rawStory['author'] ?? null,
            $rawStory['category'] ?? null,
            $rawStory['group_name'] ?? null,
            $rawStory['visibility'] ?? null,
            isset($rawStory['priority']) ? (int)$rawStory['priority'] : 0,
            isset($rawStory['peghi']) ? (int)$rawStory['peghi'] : 0,
            $diffCount
        );
    }

    /**
     * @param array $texts
     * @param int|null $txtId
     * @param string $targetLang
     * @return string|null
     */
    public function resolveText(array $texts, ?int $txtId, string $targetLang): ?string
    {
        if ($txtId === null) {
            return null;
        }

        $candidates = array_filter($texts, fn($t) => (int)$t['id_text'] === $txtId);
        if (empty($candidates)) {
            return null;
        }

        // 1. Exact language match
        foreach ($candidates as $t) {
            if ($t['lang'] === $targetLang) {
                return $t['short_text'] ?? $t['long_text'];
            }
        }

        // 2. English fallback
        foreach ($candidates as $t) {
            if ($t['lang'] === 'en') {
                return $t['short_text'] ?? $t['long_text'];
            }
        }

        // 3. Any available
        $t = reset($candidates);
        return $t['short_text'] ?? $t['long_text'];
    }
}
